File: src/orchestration/daemon.rs
//! Daemon Process Management
//! 
//! Implements robust daemonization logic handling PID file management
//! and standard stream redirection for 24/7 background operation.

use std::fs::{self, File, OpenOptions};
use std::io::{self, Write, Read};
use std::path::{Path, PathBuf};
use std::process;
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use crossbeam_utils::CachePadded;

/// Daemon state
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DaemonState {
    /// Not yet daemonized
    Starting,
    /// Running as daemon
    Running,
    /// Shutting down
    Stopping,
    /// Stopped
    Stopped,
}

/// Daemon configuration
pub struct DaemonConfig {
    /// PID file path
    pub pid_file: PathBuf,
    /// Log file path
    pub log_file: Option<PathBuf>,
    /// Working directory
    pub work_dir: PathBuf,
    /// User to run as (if different)
    pub run_as_user: Option<String>,
    /// Group to run as (if different)
    pub run_as_group: Option<String>,
    /// Umask
    pub umask: u32,
}

impl Default for DaemonConfig {
    fn default() -> Self {
        Self {
            pid_file: PathBuf::from("/var/run/hft_bot.pid"),
            log_file: Some(PathBuf::from("/var/log/hft_bot.log")),
            work_dir: PathBuf::from("/"),
            run_as_user: None,
            run_as_group: None,
            umask: 0o022,
        }
    }
}

/// Daemon manager
pub struct DaemonManager {
    /// Current state
    state: CachePadded<AtomicU64>,
    /// PID file handle
    pid_file: CachePadded<std::sync::Mutex<Option<File>>>,
    /// Log file handle
    log_file: CachePadded<std::sync::Mutex<Option<File>>>,
    /// Process start time
    start_time_ns: CachePadded<AtomicU64>,
    /// Configuration
    config: DaemonConfig,
}

impl DaemonManager {
    /// Create a new daemon manager
    pub fn new(config: DaemonConfig) -> Self {
        Self {
            state: CachePadded::new(AtomicU64::new(DaemonState::Starting as u64)),
            pid_file: CachePadded::new(std::sync::Mutex::new(None)),
            log_file: CachePadded::new(std::sync::Mutex::new(None)),
            start_time_ns: CachePadded::new(AtomicU64::new(0)),
            config,
        }
    }

    /// Check if another instance is running
    pub fn check_existing(&self) -> Result<bool, String> {
        if self.config.pid_file.exists() {
            if let Ok(content) = fs::read_to_string(&self.config.pid_file) {
                if let Ok(pid) = content.trim().parse::<u32>() {
                    // Check if process exists
                    #[cfg(unix)]
                    {
                        use libc;
                        unsafe {
                            if libc::kill(pid as i32, 0) == 0 {
                                return Ok(true); // Process exists
                            }
                        }
                    }
                    
                    #[cfg(not(unix))]
                    {
                        // On non-Unix, assume it exists if PID file exists
                        return Ok(true);
                    }
                }
            }
        }
        Ok(false)
    }

    /// Write PID file
    fn write_pid_file(&self) -> Result<(), String> {
        // Ensure parent directory exists
        if let Some(parent) = self.config.pid_file.parent() {
            if !parent.exists() {
                fs::create_dir_all(parent).map_err(|e| format!("Failed to create PID dir: {}", e))?;
            }
        }

        let mut file = OpenOptions::new()
            .write(true)
            .create(true)
            .truncate(true)
            .open(&self.config.pid_file)
            .map_err(|e| format!("Failed to open PID file: {}", e))?;

        writeln!(file, "{}", process::id())
            .map_err(|e| format!("Failed to write PID: {}", e))?;

        *self.pid_file.lock().unwrap() = Some(file);
        Ok(())
    }

    /// Remove PID file
    fn remove_pid_file(&self) {
        drop(self.pid_file.lock().unwrap().take());
        let _ = fs::remove_file(&self.config.pid_file);
    }

    /// Setup log file
    fn setup_log_file(&self) -> Result<(), String> {
        if let Some(ref log_path) = self.config.log_file {
            if let Some(parent) = log_path.parent() {
                if !parent.exists() {
                    fs::create_dir_all(parent).map_err(|e| format!("Failed to create log dir: {}", e))?;
                }
            }

            let file = OpenOptions::new()
                .write(true)
                .create(true)
                .append(true)
                .open(log_path)
                .map_err(|e| format!("Failed to open log file: {}", e))?;

            *self.log_file.lock().unwrap() = Some(file);
        }
        Ok(())
    }

    /// Redirect standard streams
    fn redirect_streams(&self) -> Result<(), String> {
        #[cfg(unix)]
        {
            use std::os::unix::io::AsRawFd;
            
            // Open /dev/null for stdin
            let null = OpenOptions::new()
                .read(true)
                .open("/dev/null")
                .map_err(|e| format!("Failed to open /dev/null: {}", e))?;

            unsafe {
                libc::dup2(null.as_raw_fd(), 0);
            }

            // Redirect stdout/stderr to log file if configured
            if let Some(ref mut log_file) = *self.log_file.lock().unwrap() {
                unsafe {
                    libc::dup2(log_file.as_raw_fd(), 1);
                    libc::dup2(log_file.as_raw_fd(), 2);
                }
            }
        }

        Ok(())
    }

    /// Daemonize the process
    pub fn daemonize(&self) -> Result<(), String> {
        // Check for existing instance
        if self.check_existing()? {
            return Err("Another instance is already running".to_string());
        }

        #[cfg(unix)]
        {
            // First fork
            unsafe {
                match libc::fork() {
                    -1 => return Err("First fork failed".to_string()),
                    0 => {
                        // Child continues
                    }
                    _ => {
                        // Parent exits
                        process::exit(0);
                    }
                }
            }

            // Create new session
            unsafe {
                if libc::setsid() == -1 {
                    return Err("setsid failed".to_string());
                }
            }

            // Ignore SIGCHLD
            unsafe {
                libc::signal(libc::SIGCHLD, libc::SIG_IGN);
            }

            // Second fork (optional, prevents zombie processes)
            unsafe {
                match libc::fork() {
                    -1 => return Err("Second fork failed".to_string()),
                    0 => {
                        // Grandchild continues
                    }
                    _ => {
                        // First child exits
                        process::exit(0);
                    }
                }
            }

            // Change working directory
            let _ = std::env::set_current_dir(&self.config.work_dir);

            // Set umask
            unsafe {
                libc::umask(self.config.umask as libc::mode_t);
            }
        }

        // Setup files
        self.setup_log_file()?;
        self.redirect_streams()?;
        self.write_pid_file()?;

        // Record start time
        self.start_time_ns.store(get_timestamp_ns(), Ordering::Relaxed);

        // Update state
        self.state.store(DaemonState::Running as u64, Ordering::Relaxed);

        Ok(())
    }

    /// Run in foreground (no daemonization)
    pub fn run_foreground(&self) -> Result<(), String> {
        self.setup_log_file()?;
        self.write_pid_file()?;
        self.start_time_ns.store(get_timestamp_ns(), Ordering::Relaxed);
        self.state.store(DaemonState::Running as u64, Ordering::Relaxed);
        Ok(())
    }

    /// Get current state
    pub fn get_state(&self) -> DaemonState {
        match self.state.load(Ordering::Relaxed) {
            0 => DaemonState::Starting,
            1 => DaemonState::Running,
            2 => DaemonState::Stopping,
            _ => DaemonState::Stopped,
        }
    }

    /// Set state to stopping
    pub fn set_stopping(&self) {
        self.state.store(DaemonState::Stopping as u64, Ordering::Relaxed);
    }

    /// Cleanup on shutdown
    pub fn cleanup(&self) {
        self.remove_pid_file();
        self.state.store(DaemonState::Stopped as u64, Ordering::Relaxed);
    }

    /// Get uptime in seconds
    pub fn get_uptime_secs(&self) -> u64 {
        let start = self.start_time_ns.load(Ordering::Relaxed);
        if start == 0 {
            return 0;
        }
        (get_timestamp_ns() - start) / 1_000_000_000
    }

    /// Get PID
    pub fn get_pid(&self) -> u32 {
        process::id()
    }

    /// Write to log file
    pub fn log(&self, message: &str) {
        if let Some(ref mut file) = *self.log_file.lock().unwrap() {
            let timestamp = get_timestamp_ns();
            let _ = writeln!(file, "[{}] {}", timestamp, message);
            let _ = file.flush();
        }
    }
}

impl Drop for DaemonManager {
    fn drop(&mut self) {
        self.remove_pid_file();
    }
}

#[inline]
fn get_timestamp_ns() -> u64 {
    use std::time::{SystemTime, UNIX_EPOCH};
    SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_nanos() as u64
}

// Stub libc for non-Unix
#[cfg(not(unix))]
mod libc {
    pub type mode_t = u32;
    
    pub const SIGCHLD: i32 = 17;
    pub const SIG_IGN: usize = 1;
    
    pub unsafe fn fork() -> i32 { -1 }
    pub unsafe fn setsid() -> i32 { -1 }
    pub unsafe fn signal(_sig: i32, _handler: usize) -> usize { 0 }
    pub unsafe fn umask(_mask: mode_t) -> mode_t { 0 }
    pub unsafe fn kill(_pid: i32, _sig: i32) -> i32 { -1 }
    pub unsafe fn dup2(_old: i32, _new: i32) -> i32 { -1 }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_daemon_config_default() {
        let config = DaemonConfig::default();
        assert_eq!(config.pid_file, PathBuf::from("/var/run/hft_bot.pid"));
    }

    #[test]
    fn test_daemon_state() {
        let config = DaemonConfig::default();
        let daemon = DaemonManager::new(config);
        
        assert_eq!(daemon.get_state(), DaemonState::Starting);
    }
}

File: python/orchestration/daemon.py
"""
Python Daemonizer.
Handles PID files, signal trapping (SIGTERM), and backgrounding.
Ensures Ray cluster and Nautilus kernel run detached with strict process control.
"""

import os
import sys
import signal
import atexit
import logging
from typing import Optional, Callable, Dict, Any
from dataclasses import dataclass
import subprocess
import time

logger = logging.getLogger(__name__)


@dataclass
class ProcessHandle:
    """Handle for a managed subprocess."""
    name: str
    pid: int
    process: subprocess.Popen
    started_at: float
    should_restart: bool = True


class PythonDaemon:
    """
    Robust Python daemonizer for HFT systems.
    Manages PID files, signal handling, and process lifecycle.
    """

    def __init__(
        self,
        pid_file: str = "/tmp/hft_bot.pid",
        log_file: str = "/tmp/hft_bot.log",
        work_dir: str = "/workspace",
    ):
        self.pid_file = pid_file
        self.log_file = log_file
        self.work_dir = work_dir

        self._processes: Dict[str, ProcessHandle] = {}
        self._running = False
        self._shutdown_callbacks: list = []
        self._original_sigterm = None
        self._original_sigint = None

    def daemonize(self) -> bool:
        """
        Daemonize the current process.

        Returns:
            True if successful
        """
        try:
            # First fork
            pid = os.fork()
            if pid > 0:
                sys.exit(0)  # Exit parent

            # Create new session
            os.setsid()

            # Second fork (prevent session leader from acquiring terminal)
            pid = os.fork()
            if pid > 0:
                sys.exit(0)

            # Change working directory
            os.chdir(self.work_dir)

            # Redirect standard file descriptors
            self._redirect_io()

            # Write PID file
            self._write_pid_file()

            # Setup signal handlers
            self._setup_signal_handlers()

            # Register cleanup
            atexit.register(self._cleanup)

            self._running = True
            logger.info(f"Daemon started with PID {os.getpid()}")
            return True

        except Exception as e:
            logger.error(f"Failed to daemonize: {e}")
            return False

    def _redirect_io(self):
        """Redirect stdin/stdout/stderr to log file."""
        # Open log file
        log_fd = os.open(self.log_file, os.O_CREAT | os.O_WRONLY | os.O_APPEND, 0o644)

        # Redirect stdin from /dev/null
        dev_null_fd = os.open(os.devnull, os.O_RDONLY)
        os.dup2(dev_null_fd, sys.stdin.fileno())
        os.close(dev_null_fd)

        # Redirect stdout and stderr to log file
        os.dup2(log_fd, sys.stdout.fileno())
        os.dup2(log_fd, sys.stderr.fileno())
        os.close(log_fd)

    def _write_pid_file(self):
        """Write PID file."""
        pid = os.getpid()
        with open(self.pid_file, 'w') as f:
            f.write(str(pid))
        logger.info(f"PID file written: {self.pid_file}")

    def _remove_pid_file(self):
        """Remove PID file."""
        try:
            if os.path.exists(self.pid_file):
                os.remove(self.pid_file)
                logger.debug("PID file removed")
        except Exception as e:
            logger.error(f"Failed to remove PID file: {e}")

    def _setup_signal_handlers(self):
        """Setup signal handlers for graceful shutdown."""
        self._original_sigterm = signal.signal(signal.SIGTERM, self._handle_signal)
        self._original_sigint = signal.signal(signal.SIGINT, self._handle_signal)
        signal.signal(signal.SIGHUP, self._handle_signal)

    def _handle_signal(self, signum, frame):
        """Handle incoming signals."""
        sig_name = signal.Signals(signum).name
        logger.info(f"Received signal: {sig_name}")

        if signum in (signal.SIGTERM, signal.SIGINT):
            self.shutdown()

    def start_process(
        self,
        name: str,
        command: list,
        should_restart: bool = True,
        env: Optional[Dict] = None,
    ) -> Optional[ProcessHandle]:
        """
        Start a managed subprocess.

        Args:
            name: Process name
            command: Command to execute
            should_restart: Whether to restart on failure
            env: Environment variables

        Returns:
            ProcessHandle or None
        """
        try:
            process_env = os.environ.copy()
            if env:
                process_env.update(env)

            proc = subprocess.Popen(
                command,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                env=process_env,
                cwd=self.work_dir,
            )

            handle = ProcessHandle(
                name=name,
                pid=proc.pid,
                process=proc,
                started_at=time.time(),
                should_restart=should_restart,
            )

            self._processes[name] = handle
            logger.info(f"Started process '{name}' with PID {proc.pid}")

            return handle

        except Exception as e:
            logger.error(f"Failed to start process '{name}': {e}")
            return None

    def stop_process(self, name: str, timeout: float = 5.0) -> bool:
        """
        Stop a managed process.

        Args:
            name: Process name
            timeout: Seconds to wait for graceful shutdown

        Returns:
            True if stopped successfully
        """
        if name not in self._processes:
            return False

        handle = self._processes[name]

        try:
            # Send SIGTERM
            handle.process.terminate()

            # Wait for graceful shutdown
            try:
                handle.process.wait(timeout=timeout)
            except subprocess.TimeoutExpired:
                # Force kill
                handle.process.kill()
                handle.process.wait()

            logger.info(f"Stopped process '{name}' (PID {handle.pid})")
            del self._processes[name]
            return True

        except Exception as e:
            logger.error(f"Failed to stop process '{name}': {e}")
            return False

    def monitor_processes(self) -> Dict[str, bool]:
        """
        Monitor all managed processes and restart if needed.

        Returns:
            Dict of process name -> running status
        """
        status = {}

        for name, handle in list(self._processes.items()):
            poll_result = handle.process.poll()

            if poll_result is None:
                # Still running
                status[name] = True
            else:
                # Process exited
                status[name] = False
                logger.warning(f"Process '{name}' exited with code {poll_result}")

                if handle.should_restart and self._running:
                    logger.info(f"Restarting process '{name}'...")
                    # Would need to store command to restart
                    # For now, just mark as not running

        return status

    def register_shutdown_callback(self, callback: Callable):
        """Register a callback to run on shutdown."""
        self._shutdown_callbacks.append(callback)

    def shutdown(self):
        """Gracefully shutdown all processes."""
        logger.info("Initiating graceful shutdown...")
        self._running = False

        # Run shutdown callbacks
        for callback in reversed(self._shutdown_callbacks):
            try:
                callback()
            except Exception as e:
                logger.error(f"Shutdown callback error: {e}")

        # Stop all processes
        for name in list(self._processes.keys()):
            self.stop_process(name, timeout=3.0)

        # Cleanup
        self._cleanup()

    def _cleanup(self):
        """Cleanup resources."""
        self._remove_pid_file()
        logger.info("Cleanup complete")

    def is_running(self) -> bool:
        """Check if daemon is running."""
        return self._running

    @staticmethod
    def check_existing_pid(pid_file: str) -> Optional[int]:
        """Check if there's an existing daemon running."""
        if not os.path.exists(pid_file):
            return None

        try:
            with open(pid_file, 'r') as f:
                pid = int(f.read().strip())

            # Check if process exists
            os.kill(pid, 0)
            return pid

        except (ValueError, ProcessLookupError, PermissionError):
            # Stale PID file
            if os.path.exists(pid_file):
                os.remove(pid_file)
            return None


# Module singleton
_daemon: Optional[PythonDaemon] = None


def get_daemon(
    pid_file: str = "/tmp/hft_bot.pid",
    log_file: str = "/tmp/hft_bot.log",
) -> PythonDaemon:
    """Get or create the daemon singleton."""
    global _daemon
    if _daemon is None:
        _daemon = PythonDaemon(pid_file=pid_file, log_file=log_file)
    return _daemon


async def shutdown_daemon_module():
    """Shutdown the daemon module."""
    global _daemon
    if _daemon:
        _daemon.shutdown()
        _daemon = None
